src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs

```diff
@@ -322,22 +322,22 @@ public ProjectInfo WithParseOptions(ParseOptions? parseOptions)
             => With(parseOptions: parseOptions);
 
         public ProjectInfo WithDocuments(IEnumerable<DocumentInfo>? documents)
-            => With(documents: documents.AsBoxedImmutableArrayWithNonNullItems() ?? throw new ArgumentNullException(nameof(documents)));
+            => With(documents: documents.ToBoxedImmutableArrayWithDistinctNonNullItems(nameof(documents)));
 
         public ProjectInfo WithAdditionalDocuments(IEnumerable<DocumentInfo>? additionalDocuments)
-            => With(additionalDocuments: additionalDocuments.AsBoxedImmutableArrayWithNonNullItems() ?? throw new ArgumentNullException(nameof(additionalDocuments)));
+            => With(additionalDocuments: additionalDocuments.ToBoxedImmutableArrayWithDistinctNonNullItems(nameof(additionalDocuments)));
 
         public ProjectInfo WithAnalyzerConfigDocuments(IEnumerable<DocumentInfo>? analyzerConfigDocuments)
-            => With(analyzerConfigDocuments: analyzerConfigDocuments.AsBoxedImmutableArrayWithNonNullItems() ?? throw new ArgumentNullException(nameof(analyzerConfigDocuments)));
+            => With(analyzerConfigDocuments: analyzerConfigDocuments.ToBoxedImmutableArrayWithDistinctNonNullItems(nameof(analyzerConfigDocuments)));
```

Comment:
It feels a bit funky that these are all losing the throw new ArgumentNullException() checks and that's now being handled internally. It means this now behaves somewhat differently than the rest of our To* methods, and requires you to guess which class this is coming from. Since it's an extension method, the "this is from PublicContract" gets lost which is somewhat key IMHO.

Reply:
I'm not sure I understand. `ToBoxedImmutableArrayWithDistinctNonNullItems` throws for exactly the same inputs as `AsBoxedImmutableArrayWithNonNullItems() ?? throw new ArgumentNullException(nameof(analyzerConfigDocuments))` did.

Reply:
Oh, I see what you mean. It's the fact that this is an extension method. Makes sense, I can make it a static method.